### PR TITLE
Abstract importnames

### PR DESCRIPTION
**Pull request by si-23**

### CODE REVIEW

**Review by Synss — APPROVED** [review]

File: py_import_cycles/main.py [hunk 1]
```diff
@@ -89,36 +89,74 @@
...
+
+    def __gt__(self, other: object) -> bool:
+        if not isinstance(other, ModuleName):
+            raise TypeError()
+        return self._name > other._name
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, ModuleName):
+            raise TypeError()
+        return self._name == other._name
```
Comment (Synss) [inline py_import_cycles/main.py:115]: I think
```
ModuleName("a", "b", "c") == "a.b.c"
ModuleName("a", "b", "c") == ("a", "b", "c")
ModuleName("a.b", "c") == ModuleName("a.b.c")
ModuleName("a.b", "c") == ModuleName("a", "b", "c")
```
could be equal as well.

File: py_import_cycles/main.py [hunk 2]
```diff
@@ -89,36 +89,74 @@
...
+    @property
+    def parent(self) -> ModuleName:
+        return ModuleName(*self._name.rsplit(".", 1)[:-1])
+
+    @property
+    def parents(self) -> Sequence[ModuleName]:
+        parts = self._name.split(".")
+        return [ModuleName(".".join(parts[:idx])) for idx in range(1, len(parts))][::-1]
+
+    def joinname(self, name: str) -> ModuleName:
```
Comment (Synss) [inline py_import_cycles/main.py:134]: `relative_to` may be useful as well.

I think one way to canonicalize this `ModuleName` should be required and that would be `Path.resolve()` in pathlib.

File: py_import_cycles/main.py [hunk 3]
```diff
@@ -89,36 +89,74 @@
...
 
 Module = Union[RegularPackage, NamespacePackage, PyModule]
 
 
 def _make_module_from_name(
-    mapping: Mapping[str, str], project_path: Path, module_name: str
+    mapping: Mapping[str, str], project_path: Path, module_name: ModuleName
 ) -> Optional[Module]:
-    parts = list(_make_module_parts(module_name))
+    parts = list(module_name.parts)
```
Comment (Synss) [inline py_import_cycles/main.py]: This is reverting to `List[str]` to make the new abstraction class useful. I would rather extend the implementation of the class to handle, at least, simple things such as appending/prepending. `pathlib.Path` does that using operator overloading `Path("x") / "y"`. That would be an idea here as well.

File: py_import_cycles/main.py [hunk 4]
```diff
@@ -289,35 +323,29 @@ def _get_name_prefix(self, import_from_stmt: ast.ImportFrom) -> str:
...
         #   -> Python module/package: ../BASE_PARENT/a/b{.py,/}
         if import_from_stmt.level == 0:
-            return import_from_stmt.module if import_from_stmt.module else ""
-
-        module_name_parts = list(_make_module_parts(self._base_module.name))[
-            : -import_from_stmt.level
-        ]
-        if import_from_stmt.module:
-            module_name_parts += list(_make_module_parts(import_from_stmt.module))
+            return ModuleName(import_from_stmt.module if import_from_stmt.module else "")
```
Comment (Synss) [inline py_import_cycles/main.py:324]: That may indeed create a `ModuleName("")` that, I believe, raises IndexError in a few cases as mentioned above.

Here again, we really need unit tests.

File: py_import_cycles/main.py [hunk 5]
```diff
@@ -289,35 +323,29 @@ def _get_name_prefix(self, import_from_stmt: ast.ImportFrom) -> str:
...
-        return _make_module_name(*module_name_parts)
+        parent = self._base_module.name.parents[import_from_stmt.level - 1]
+        return parent.joinname(import_from_stmt.module) if import_from_stmt.module else parent
 
     # -----helper-----
 
-    def _get_module(self, name: str) -> Optional[Module]:
-        if self._is_builtin_or_stdlib(name):
+    def _get_module(self, module_name: ModuleName) -> Optional[Module]:
+        if self._is_builtin_or_stdlib(module_name.name):
```
Comment (Synss) [inline py_import_cycles/main.py:332]: `is_builtin_or_stdlib` could be defined on `ModuleName` as well.

At least, I believe it should be a free function (or two, one for builtin and one for stdlib).

File: py_import_cycles/main.py [hunk 6]
```diff
@@ -273,12 +309,10 @@ def _get_modules_of_import_from_stmt(
...
             # 3 -> ../BASE/a/b/c{.py,/}
             # 4 -> ../BASE_PARENT/a/b/c{.py,/}
-            if this_imported_module := self._get_module(
-                _make_module_name(module_name_prefix, alias.name)
-            ):
+            if this_imported_module := self._get_module(module_name_prefix.joinname(alias.name)):
                 yield this_imported_module
 
-    def _get_name_prefix(self, import_from_stmt: ast.ImportFrom) -> str:
+    def _get_name_prefix(self, import_from_stmt: ast.ImportFrom) -> ModuleName:
```
Comment (Synss) [inline py_import_cycles/main.py:313]: should most likely be a free function

**Review by Synss — CHANGES_REQUESTED** [review]

File: py_import_cycles/main.py [hunk 7]
```diff
@@ -153,66 +153,66 @@ class PyModule(NamedTuple):
...
-        if value in parts:
-            parts = [key] + parts
-            break
+@dataclass(frozen=True)
+class ModuleFactory:
+    _mapping: Mapping[str, str]
+    _project_path: Path
 
-    module_path = project_path.joinpath(Path(*parts))
+    def from_name(self, module_name: ModuleName) -> None | Module:
```
Comment (Synss) [inline py_import_cycles/main.py:161]: It seems to me `from_name()` and `from_path()` do not have much to do with one another. I also find the name confusing. I would expect a `from_` function would be an alternative constructor such as
```
class K:
    @classmethod
    def from_x(cls, x: X) -> K:
        ...
```
there are example of this pattern in the stdlib as well.

Before this patch, both functions would return `Module` and now, one returns `Module | None` and the other `RegularPackage | PyModule`.

I think I do not understand this part of the code.

File: py_import_cycles/main.py [hunk 8]
```diff
@@ -330,15 +328,11 @@ def _get_name_prefix(self, import_from_stmt: ast.ImportFrom) -> ModuleName:
 
     # -----helper-----
 
-    def _get_module(self, module_name: ModuleName) -> Optional[Module]:
+    def _get_module(self, module_name: ModuleName) -> None | Module:
```
Comment (Synss) [inline py_import_cycles/main.py:331]: more common is, I believe, `Module | None`. `Module` is the important type. That is also what `pyupgrade` does.

**Review by Synss — CHANGES_REQUESTED** [review]

**Comment by si-23** [comment]
Thx for your comments which are very useful and convincing.

**Review by Synss — APPROVED** [review]

**Review by Synss — APPROVED** [review]

**Review by Synss — APPROVED** [review]

**Pull request merged**